### PR TITLE
SQL: Implement FIRST/LAST aggregate functions

### PR DESCRIPTION
**Pull request by matriv**
FIRST and LAST can be used with one argument and work similarly to MIN
and MAX but they are implemented using a Top Hits aggregation and
therefore can also operate on keyword fields. When a second argument is
provided then they return the first/last value of the first arg when its
values are ordered ascending/descending (respectively) by the values of
the second argument. Currently because of the usage of a Top Hits
aggregation FIRST and LAST cannot be used in the HAVING clause of a
GROUP BY query to filter on the results of the aggregation.

Closes: #35639

### CODE REVIEW

File: docs/reference/sql/functions/aggs.asciidoc [hunk 1]
```diff
@@ -113,6 +113,129 @@ Returns the total number of _distinct non-null_ values in input values.
 include-tagged::{sql-specs}/docs.csv-spec[aggCountDistinct]
 --------------------------------------------------
 
+[[sql-functions-aggs-first]]
+===== `FIRST/FIRST_VALUE`
+
+.Synopsis:
+[source, sql]
+--------------------------------------------------
```
Comment (costin) [inline docs/reference/sql/functions/aggs.asciidoc]: I think for optional arguments, it would be nice to add an example with `FIRST` just with the first argument.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 2]
```diff
@@ -113,6 +113,129 @@ Returns the total number of _distinct non-null_ values in input values.
...
+FIRST(field_name<1>, sort_by_field_name<2>)
```
Comment (costin) [inline docs/reference/sql/functions/aggs.asciidoc]: sort_by_field_name -> ordering field

File: docs/reference/sql/functions/aggs.asciidoc [hunk 3]
```diff
@@ -113,6 +113,129 @@ Returns the total number of _distinct non-null_ values in input values.
...
+--------------------------------------------------
+
+*Input*:
+
+<1> a field name
```
Comment (costin) [inline docs/reference/sql/functions/aggs.asciidoc]: <1> target field for the aggregation
<2> optional field used for ordering

File: docs/reference/sql/functions/aggs.asciidoc [hunk 4]
```diff
@@ -113,6 +113,129 @@ Returns the total number of _distinct non-null_ values in input values.
...
+.Description:
+
+When only one argument is provided it returns the first **non-NULL** value across input values in the field
+`field_name`. It will return **NULL** only if all values in `field_name` are null. When a second argument
+is provided then it returns the first **non-NULL** value across input values in the field `field_name` ordered
+ascending by the **non-NULL** values of `sort_by_field_name`. E.g.:
+
+[cols="<,<"]
+|===
+s|a|b
```
Comment (costin) [inline docs/reference/sql/functions/aggs.asciidoc]: Please use indentation to have the table columns aligned.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 5]
```diff
@@ -113,6 +113,129 @@ Returns the total number of _distinct non-null_ values in input values.
 include-tagged::{sql-specs}/docs.csv-spec[aggCountDistinct]
 --------------------------------------------------
 
+[[sql-functions-aggs-first]]
+===== `FIRST/FIRST_VALUE`
```
Comment (costin) [inline docs/reference/sql/functions/aggs.asciidoc]: Additionally please add an example for `FIRST_VALUE`.

File: x-pack/plugin/sql/qa/src/main/resources/docs.csv-spec [hunk 6]
```diff
@@ -998,6 +1004,55 @@ SELECT COUNT(DISTINCT hire_date) unique_hires, COUNT(hire_date) AS hires FROM em
 // end::aggCountDistinct
 ;
 
+firstWithOneArg
+schema::FIRST(first_name):s
+// tag::firstWithOneArg
+SELECT FIRST(first_name) FROM emp;
```
Comment (costin) [inline x-pack/plugin/sql/qa/src/main/resources/docs.csv-spec]: I think the examples would be more meaningful along with a GROUP BY.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java [hunk 7]
```diff
@@ -407,13 +418,17 @@ private static boolean checkGroupByHavingHasOnlyAggs(Expression e, Node<?> sourc
...
             }
             return true;
 
         } else if (e instanceof Score) {
-            // Score can't be used for having
-            missing.put(e, source);
+            // Score can't be used in having
+            unsupported.add(e);
+            return true;
+        } else if (e instanceof TopHits) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java]: I believe this will not catch `MAX(keyword)` pass by.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (field().foldable()) {
+            return new TypeResolution(format(null, "First argument of [{}] must be a table column, found constant [{}]",
+                functionName(),
+                Expressions.name(field())));
+        }
+        try {
+            ((FieldAttribute) field()).exactAttribute();
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: Use `FieldAttribute.isExact` instead

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        if (sortField() != null) {
+            if (sortField().foldable()) {
+                return new TypeResolution(format(null, "Second argument of [{}] must be a table column, found constant [{}]",
+                    functionName(),
+                    Expressions.name(sortField())));
+            }
+            try {
+                ((FieldAttribute) sortField()).exactAttribute();
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: `FieldAttribute.isExact`

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.sql.type.DataType;
+
+import java.util.Collections;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+
+public abstract class TopHits extends AggregateFunction {
+
+    TopHits(Source source, Expression field, Expression sortField) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: Create a constructor without the sortField and use that inside `First` and `Last` - this will avoid having to do `First(source, field, null)`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 11]
```diff
@@ -622,6 +627,38 @@ protected LogicalPlan rule(LogicalPlan e) {
         }
     }
 
+    static class ReplaceMinMaxWithTopHits extends OptimizerRule<Aggregate> {
+
+        @Override
+        protected LogicalPlan rule(Aggregate aggregate) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: When rewriting an expression, its id needs taking into account:
`SELECT MAX(keyword) AS m FROM x GROUP BY c ORDER BY m`
max has id: 1, get written to First with id: 5 and ORDER BY m (which points to id 1) becomes invalid since that expression id is not available anymore.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        super(source, field, sortField != null ? Collections.singletonList(sortField) : Collections.emptyList());
+    }
+
+    public Expression sortField() {
+        return parameters().isEmpty() ? null : parameters().get(0);
+    }
+
+    public DataType sortFieldDataType() {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: I don't think this method is needed - one can always say `sortField().dataType()`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+
+public abstract class TopHits extends AggregateFunction {
+
+    TopHits(Source source, Expression field, Expression sortField) {
+        super(source, field, sortField != null ? Collections.singletonList(sortField) : Collections.emptyList());
+    }
+
+    public Expression sortField() {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: Any reason why sort is used instead of order/orderby?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/TopHitsAgg.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        this.sortOrder = sortOrder;
+        this.fieldDataType = fieldDataType;
+        this.sortFieldDataType = sortFieldDataType;
+    }
+
+    @Override
+    AggregationBuilder toBuilder() {
+        // Sort missing values (NULLs) as last to get the first/last non-null value
+        List<SortBuilder<?>> sortBuilderList;
+        if (sortField != null) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/TopHitsAgg.java]: The code here is redundant since the field ordering code is duplicated.
```
if (sortField!=null) {
  sort.add(fieldSortBuilder(sortField);
}

sort.add(fieldSortBuilder(fieldName());
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/TopHitsAgg.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+            sortBuilderList = Arrays.asList(
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/TopHitsAgg.java]: To increase readability please use one method per line.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java [hunk 16]
```diff
@@ -240,4 +243,8 @@ public static DataType fromTypeName(String esType) {
             return DataType.UNSUPPORTED;
         }
     }
+
+    public String format() {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: Not okay conceptually and implementation wise since the base class is aware of its child type.
Make format() return DocValueFieldsContext by default and override DateTime and Date to return DATE_PARSE_FORMA - basic overriding.

**Review by costin — CHANGES_REQUESTED** [review]
Looks good overall however there some things that need improving.
In particular the rewrite optimization I don't think is complete and probably needs a follow-up PR.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java (hunk 12 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: I did that to avoid the caller checking if the `sortField()` is null.

File: docs/reference/sql/functions/aggs.asciidoc (hunk 1 above)
Comment (matriv) [reply docs/reference/sql/functions/aggs.asciidoc]: You mean here in the synopsis? because we have an example here: https://github.com/elastic/elasticsearch/pull/37936/files/1be7ba194c950ab175ca0e35d927ab0b54306be2#diff-f71f0c2cd27202241aec808b033b8510R166

File: x-pack/plugin/sql/qa/src/main/resources/docs.csv-spec (hunk 6 above)
Comment (matriv) [reply x-pack/plugin/sql/qa/src/main/resources/docs.csv-spec]: Generally we don't have group by for min/max/avg/kurtosis etc. examples

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java (hunk 13 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: Not really, can change.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java (hunk 16 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: But this is an enum, I don't get it.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 17]
```diff
@@ -161,6 +284,9 @@ Returns the minimum value across input values in the field `field_name`.
 include-tagged::{sql-specs}/docs.csv-spec[aggMin]
 --------------------------------------------------
 
+[NOTE]
+`MIN` on a field of type <<text, `text`>> or <<keyword, `keyword`>> is translated into <<sql-functions-aggs-first>>.
```
Comment (astefan) [inline docs/reference/sql/functions/aggs.asciidoc]: And I'm assuming MIN/MAX on keywords cannot be used in `HAVING`s...

**Review by astefan — COMMENTED** [review]
Looking good. Left few minor comments.
Also, I would like to see tests for all values NULL. Something like `SELECT FIRST(whatever) FROM bla WHERE whatever IS NULL GROUP BY X`.
Also, tests that catch the error generated when FIRST/LAST/MIN/MAX is used in HAVING.

File: docs/reference/sql/functions/aggs.asciidoc (hunk 1 above)
Comment (costin) [reply docs/reference/sql/functions/aggs.asciidoc]: Right, for synopsis.

File: x-pack/plugin/sql/qa/src/main/resources/docs.csv-spec (hunk 6 above)
Comment (costin) [reply x-pack/plugin/sql/qa/src/main/resources/docs.csv-spec]: We have at least one query with GROUP BY for each agg - the idea is to underline that this is in fact an agg (which might not be obvious when using implicit grouping).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java (hunk 12 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: All expressions return only one datatype - of their result, adding a new method is confusing. It's simple to do the null check - in fact, it's is mandatory for sorting since one cannot just add a null, null directive (if the sort field would be null).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java (hunk 16 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: Let me rephrase this - a dataType can have different formats. Currently we store that not in the `DataType` but rather inside the `DateEsField` as it can vary from field to field.
Ironically due to our usage of field_caps this information is not accessible anymore however it might be in the future.
DateEsField allows a custom format to be specified though I'm not sure this is something we need to use necessarily.

If the format could be used on `DateEsField` I would opt for that, if not, it can stay as is in the `DataType` but the two classes need to be aligned so the format is defined and stored only in one place.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 18]
```diff
@@ -113,6 +113,129 @@ Returns the total number of _distinct non-null_ values in input values.
...
+*Input*:
+
+<1> a field name
+<2> a field name; optional
+
+*Output*: same type as the input
+
+.Description:
+
+When only one argument is provided it returns the first **non-NULL** value across input values in the field
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/aggs.asciidoc]: The first non-null value is returned regardless of the presence of the optional argument.
I would rephrase this to "Returns the first non-NULL value (if it exists) of the input column sorted by the sort_column. If sort_column is missing, the field_name column is used for sorting`. or something along those lines.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 19]
```diff
@@ -113,6 +113,129 @@ Returns the total number of _distinct non-null_ values in input values.
...
+include-tagged::{sql-specs}/docs.csv-spec[firstWithOneArg]
+-----------------------------------------------------------
+
+["source","sql",subs="attributes,macros"]
+-----------------------------------------------------------
+include-tagged::{sql-specs}/docs.csv-spec[firstWithTwoArgs]
+-----------------------------------------------------------
+
+[NOTE]
+`FIRST` cannot be used in to create a filter in a `HAVING` clause of a `GROUP BY` query.
```
[the remark this answers is not in the document]
Comment (costin) [reply docs/reference/sql/functions/aggs.asciidoc]: `FIRST` cannot be used in a `HAVING` clause (no need to specify `GROUP BY` as that's understood and also might be implicit).

**Review by costin — COMMENTED** [review]
Left a new batch of comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java (hunk 10 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: Cannot do as if we have 2 constructors the FunctionRegistry with the current infrastructure there gets confused on which method to use. We have only one call with the `null` arg in the Optimizer so I'd rather leave it as is.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java (hunk 8 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/TopHits.java]: I don't get this. If the field is not exact then we shouldn't throw an exception unless it doesn't have a KEYWORD exact field. to check that you need to call to get the exact field and if no KEYWORD is present you get the Mapping exception from `TextEsField.getExactField()`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java (hunk 16 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: The `getFormats()` on `DateEsField` is currently not used anywhere so I'll remove it from there to avoid any confusion.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java (hunk 7 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java]: You're right as the Verifier is checking before the optimisations.

File: docs/reference/sql/functions/aggs.asciidoc (hunk 17 above)
Comment (matriv) [reply docs/reference/sql/functions/aggs.asciidoc]: Yes, that's why I made the link here. Do you think I should add the info about HAVING here too?

Comment (astefan) [reply docs/reference/sql/functions/aggs.asciidoc]: Yes, please.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/TopHitsAggExtractor.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+import org.elasticsearch.xpack.sql.SqlIllegalArgumentException;
+import org.elasticsearch.xpack.sql.type.DataType;
+import org.elasticsearch.xpack.sql.util.DateUtils;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class TopHitsAggExtractor implements BucketExtractor {
+
+    static final String NAME = "th";
```
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/TopHitsAggExtractor.java]: Nope, is used in the `BucketExtractors` for the registration.

**Comment by matriv** [comment]
@costin @astefan Thank you for the review. Addressed comments.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 21]
```diff
@@ -113,6 +113,185 @@ Returns the total number of _distinct non-null_ values in input values.
...
+===== `FIRST/FIRST_VALUE`
+
+.Synopsis:
+[source, sql]
+--------------------------------------------------
+FIRST(field_name<1>)
+FIRST_VALUE(field_name<1>)
+
+FIRST(field_name<1>, ordering_field_name<2>)
+FIRST_VALUE(field_name<1>, ordering_field_name<2>)
```
Comment (astefan) [inline docs/reference/sql/functions/aggs.asciidoc]: Square brackets for optional fields, please.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 22]
```diff
@@ -113,6 +113,185 @@ Returns the total number of _distinct non-null_ values in input values.
...
+--------------------------------------------------
+
+*Input*:
+
+<1> target field for the aggregation
```
Comment (astefan) [inline docs/reference/sql/functions/aggs.asciidoc]: "aggregation" refers to technical implementation in the background. My personal approach would be not to expose this, but try to explain what the field is used for from the end-user perspective. Again, personal preference.

File: docs/reference/sql/functions/aggs.asciidoc [hunk 23]
```diff
@@ -113,6 +113,185 @@ Returns the total number of _distinct non-null_ values in input values.
...
+===== `LAST/LAST_VALUE`
+
+.Synopsis:
+[source, sql]
+--------------------------------------------------
+LAST(field_name<1>)
+LAST_VALUE(field_name<1>)
+
+LAST(field_name<1>, ordering_field_name<2>)
+LAST_VALUE(field_name<1>, ordering_field_name<2>)
```
Comment (astefan) [inline docs/reference/sql/functions/aggs.asciidoc]: Square brackets for optional fields, please.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 24]
```diff
@@ -669,7 +669,7 @@ public void testTopHitsAggregationWithTwoArgs() {
             assertTrue(eqe.output().get(0).dataType() == DataType.KEYWORD);
             assertThat(eqe.queryContainer().aggs().asAggBuilder().toString().replaceAll("\\s+", ""),
                 endsWith("\"top_hits\":{\"from\":0,\"size\":1,\"version\":false,\"seq_no_primary_term\":false," +
-                    "\"explain\":false,\"docvalue_fields\":[{\"field\":\"keyword\",\"format\":\"use_field_mapping\"}]," +
+                    "\"explain\":false,\"docvalue_fields\":[{\"field\":\"keyword\"}]," +
                     "\"sort\":[{\"int\":{\"order\":\"desc\",\"missing\":\"_last\",\"unmapped_type\":\"integer\"}}," +
                     "{\"keyword\":{\"order\":\"desc\",\"missing\":\"_last\",\"unmapped_type\":\"keyword\"}}]}}}}}"));
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: I cannot wrap my head around `missing` option for top_hits in the context of first/last/min/max... It's about how `null`/missing values are treated when sorting: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-sort.html#_missing_values. Is it ok to keep `missing` with the default (`_last`), or I'm missing something, or there are no tests that have `missing: _first` scenario included...?

**Review by astefan — COMMENTED** [review]
Left some minor comments.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 24 above)
Comment (astefan) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: After I clicked "add comment" I realized these are about first/last non-NULL values. So yeah, we don't care about `missing` and we like the fact that the default is `_last`.

**Review by astefan — APPROVED** [review]

File: docs/reference/sql/functions/aggs.asciidoc (hunk 21 above)
Comment (matriv) [reply docs/reference/sql/functions/aggs.asciidoc]: But now we list both with 1 arg and 2args, do you think is still necessary?

Comment (astefan) [reply docs/reference/sql/functions/aggs.asciidoc]: If we do it like this (all versions listed - with/without the optional parameter), our documentation will be inconsistent: https://github.com/elastic/elasticsearch/blob/master/docs/reference/sql/functions/string.asciidoc#locate. If I would have to choose between the two approaches, I would go with `[ ]`.

Comment (matriv) [reply docs/reference/sql/functions/aggs.asciidoc]: In my opinion you either have one signature with the 2nd arg in square brackets or you list the 2 variants.

File: docs/reference/sql/functions/aggs.asciidoc (hunk 1 above)
Comment (astefan) [reply docs/reference/sql/functions/aggs.asciidoc]: @costin our documentation, until now, lists one version of a function in synopsis where the optional arguments are between square brackets. See https://github.com/elastic/elasticsearch/blob/master/docs/reference/sql/functions/string.asciidoc#locate and https://github.com/elastic/elasticsearch/blob/master/docs/reference/sql/functions/math.asciidoc#round (and others). Listing all versions of a function (with/without optional params) imo would look a bit crowded and not-so-clean and sleek. In this particular case of FIRST/FIRST_VALUE/LAST/LAST_VALUE means we need to list four versions of the same function in the synopsis and I personally believe it doesn't look that great.

Do you believe that, maybe, it is not clear that parameter is optional? (the square brackets and the description saying "optional")

Comment (costin) [reply docs/reference/sql/functions/aggs.asciidoc]: It's not the optional parameter but the method name that I'd like to be exemplified through a query to make the function name more obvious.
It might just be easier to create a separate section for the alias, mentioning that it's an alias for ...
This can be done in a separate PR though since it's just documentation.

File: docs/reference/sql/functions/aggs.asciidoc (hunk 21 above)
Comment (costin) [reply docs/reference/sql/functions/aggs.asciidoc]: I opt for the signature with the second arguments in square brackets.

File: docs/reference/sql/functions/aggs.asciidoc (hunk 22 above)
Comment (costin) [reply docs/reference/sql/functions/aggs.asciidoc]: Not necessarily - for the aggregate function (aggregation is ES terminology, aggregate is SQL).

**Review by costin — APPROVED** [review]

File: docs/reference/sql/functions/aggs.asciidoc (hunk 1 above)
Comment (matriv) [reply docs/reference/sql/functions/aggs.asciidoc]: Added here: https://github.com/elastic/elasticsearch/pull/37936/files#diff-f71f0c2cd27202241aec808b033b8510R198
